### ISSUE TITLE
find_library does not check that libraries are usable

### ISSUE THREAD

**Issue by bruce-richardson** [mesonbuild/meson#3807]
Inside find_library, there is a basic check for a library just linking using "-l<libname>". If that fails, then we loop over a set of search directories looking for a copy of the library.
```

             for suffix in suffixes:
                 for prefix in prefixes:
                     trial = os.path.join(d, prefix + libname + '.' + suffix)
                     if os.path.isfile(trial):
                          return [trial]
```

unfortunately, this never actually checks that the found library "trial" is usable.

A specific case where this causes problems  is when doing a 32-bit build on a 64-bit system. I have installed libbsd-dev package which gives me the 64-bit libraries but not the 32-bit one. When compiling a 32-bit app, and using find_library(), find_library returns the 64-bit libbsd.a file, which then causes ninja to fail the link when actually building.

One potential fix is to check the build with each file. For static archives, a --whole-archive flag will need to be used to actually have the linker check the architecture type if there are no unresolved symbols in the test c code compiled.

**Comment by bruce-richardson** [comment]
PR #3792  is an attempt to resolve this

**Comment by nirbheek** [comment]
Fixed with https://github.com/mesonbuild/meson/pull/3792. In the future, please add "Closes #3807", etc to the commit that fixes an issue in the PR so that this happens automatically.

**Comment by bruce-richardson** [comment]
Ok. I’ll try to remember, thanks.

On 6 Jul 2018, at 18:15, Nirbheek Chauhan <notifications@github.com<mailto:notifications@github.com>> wrote:


Fixed with #3792<https://github.com/mesonbuild/meson/pull/3792>. In the future, please add "Closes #3807<https://github.com/mesonbuild/meson/issues/3807>", etc to the commit that fixes an issue in the PR so that this happens automatically.

**Comment by thesamesam** [comment]
See also https://github.com/mesonbuild/meson/issues/10936.
